sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -413,29 +415,48 @@ private[spark] class HiveExternalCatalog(client: HiveClient, hadoopConf: Configu
-        val storage = if (table.tableType == EXTERNAL && table.storage.inputFormat.isEmpty) {
-          table.storage.copy(locationUri = None)
+        if (provider == "hive") {
+          val schemaFromTableProps = getSchemaFromTableProperties(table)
+          if (DataType.equalsIgnoreCaseAndNullability(schemaFromTableProps, table.schema)) {
+            table.copy(
+              schema = schemaFromTableProps,
+              provider = Some(provider),
+              partitionColumnNames = getPartitionColumnsFromTableProperties(table),
+              bucketSpec = getBucketSpecFromTableProperties(table),
+              properties = getOriginalTableProperties(table))
+          } else {
+            // Hive metastore may change the table schema, e.g. schema inference. If the table
+            // schema we read back is different from the one in table properties which was written
+            // when creating table, we should respect the table schema from hive.
+            table.copy(
+              properties = getOriginalTableProperties(table))
+          }
         } else {
-          table.storage
+          // SPARK-15269: Persisted data source tables always store the location URI as a storage
+          // property named "path" instead of standard Hive `dataLocation`, because Hive only
+          // allows directory paths as location URIs while Spark SQL data source tables also
+          // allows file paths. So the standard Hive `dataLocation` is meaningless for Spark SQL
+          // data source tables.
+          // Spark SQL may also save external data source in Hive compatible format when
+          // possible, so that these tables can be directly accessed by Hive. For these tables,
+          // `dataLocation` is still necessary. Here we also check for input format because only
+          // these Hive compatible tables set this field.
+          val storage = if (provider != "hive" && table.tableType == EXTERNAL &&
+            table.storage.inputFormat.isEmpty) {
+            table.storage.copy(locationUri = None)
+          } else {
+            table.storage
+          }
+          table.copy(
+            storage = storage,
+            schema = getSchemaFromTableProperties(table),
+            provider = Some(provider),
+            partitionColumnNames = getPartitionColumnsFromTableProperties(table),
+            bucketSpec = getBucketSpecFromTableProperties(table),
+            properties = getOriginalTableProperties(table))
         }
-        table.copy(
-          storage = storage,
-          schema = getSchemaFromTableProperties(table),
-          provider = Some(provider),
-          partitionColumnNames = getPartitionColumnsFromTableProperties(table),
-          bucketSpec = getBucketSpecFromTableProperties(table),
-          properties = getOriginalTableProperties(table))
       } getOrElse {
+        // Hive serde tables that are created at Hive side will hit this branch.
```

Comment:
It sounds like we do not have any test case to cover this scenario. Do you want me to add it?